src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -753,21 +765,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
             {
                 var index = command.Index;
                 var options = mergedCommands.Options[index];
-                Task indexTask = null;
+                var taskResult = new ClusterTransactionResult
+                {
+                    GeneratedResult = mergedCommands.Replies[index],
+                };
                 if (options.WaitForIndexesTimeout != null)
                 {
-                    indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
+                    var indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
                         options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
                             mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                }
 
-                var result = new ClusterTransactionCompletionResult
+                    var removeTask = ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(command.Options.TaskId);
```

Comment:
Why `removeTask`? What is being removed here?

Reply:
this task is held by the ClusterTransactionWaiter and its completion tells us that the cluster transaction is completed (finished in the database).
The 'remove task' is IDisposable that at his disposal removes a cluster transaction task from the ClusterTransactionWaiter.
At the end of the cluster transaction on the database, if we need to 'wait for indexing' we create a task in the ClusterTransactionWaiter and set the task to be completed only when the indexing (task) is completed.
The task won't be created if it's already in the ClusterTransactionWaiter (created somewhere else),
in that case we will get the existing task when we'll call the ClusterTransactionWaiter.CreateTask.
We will remove the task from the ClusterTransactionWaiter dictionary, in assumption that in the places that we should 'await' that task, we already got its instance before it was removed from the ClusterTransactionWaiter dict (by disposing of its 'remove task').
We remove it here because it can cause a memory leak..
Creating the task here is adding it to the ClusterTransactionWaiter dict, and we cannot be sure that it will be removed from the dict somewhere else.